### ISSUE TITLE
Weird behaviour of || with an NQPMatch capture

### ISSUE THREAD

**Issue by zoffixznet** [perl6/nqp#326]
If I use `||` to sub a value if capture is not truthy, I get a weird error about nulls... that's even in the case where the NQPMatch capture on LHS is truthy. The same construct works fine with `//`:

```
03:22		nqp: my $m := ("foo" ~~ /$<bar>=\w+/); nqp::say($m<bar> || '')
03:22	camelia	nqp-moarvm: OUTPUT«say requires a concrete string, but got null␤   at <tmp>:1  (<ephemeral file>:<mainline>)␤ from gen/moar/stage2/NQPHLL.nqp:1462  (/home/camelia/rakudo-m-inst-1/sh​are/nqp/lib/NQPHLL.moarvm:eval)␤ from gen/moar/stage2/NQPHLL.nqp:1665  (/home/camelia/rakudo-m-i…»
03:22	babydrop	nqp: my $m := ("foo" ~~ /$<bar>=\w+/); nqp::say($m<bar> // '')
03:22	camelia	nqp-moarvm: OUTPUT«foo␤»
```

**Comment by W4anD0eR96** [comment]
@zoffixznet I can't reproduce it at commit 5f25c81ea67

**Comment by zoffixznet** [comment]
Still exists on camelia, which I assume is HEAD:

```
0617     Zoffix │ nqp: my $m := ("foo" ~~ /$<bar>=\w+/); nqp::say($m<bar> || '')                                                                                                                                                                 │
0617    camelia │ nqp-moarvm: OUTPUT: «say requires a concrete string, but got null␤   at <tmp>:1  (<ephemeral file>:<mainline>)␤ from gen/moar/stage2/NQPHLL.nqp:1542  (/home/camelia/rakudo-m-inst-1/share/nqp/lib/NQPHLL.moarvm:eval)␤ from   │
                │ gen/moar/stage2/NQPHLL.nqp:1779  (/home/camelia/rakudo-m…»      
```

**Comment by W4anD0eR96** [comment]
I use the one-line below in `nqp` REPL

```
>  my $m := ("foo" ~~ /$<bar>=\w+/);say($m<bar> || '')
```
which produces `foo`

I think `nqp` REPL itself is somewhat weird.

**Comment by zoffixznet** [comment]
In REPL, I also notice running the line again gives a different error:

```
$ nqp
> my $m := ("foo" ~~ /$<bar>=\w+/); nqp::say($m<bar> || '')
say requires a concrete string, but got null
> my $m := ("foo" ~~ /$<bar>=\w+/); nqp::say($m<bar> || '')
Too many positionals passed; expected 1 argument but got 2
> 
```

**Comment by W4anD0eR96** [comment]
Yes, I reproduced it.

**Comment by W4anD0eR96** [comment]
It's fix by chance when I fight with #332 , now

`my $m := ("foo" ~~ /$<bar>=\w+/); say($m<bar> || '')` produces right answer.

However, if you run twice, the other problem remains

**Comment by pmurias** [comment]
8a06289378bf6eee49106793a48801bae8e09c6a fixed this
